### PR TITLE
285 use priority in esdl instead of marginal costs

### PR DESCRIPTION
**Pull request by Jaimepatt**
PR addressing issue https://github.com/Project-OMOTES/simulator-core/issues/285

In esdl, it is now possible to assign control strategies to a producer asset. One of these strategies is priority based. This PR implements this functionality into the simulator:
- If at least one producer has a priority based strategy with a priority value assign, the controller now sorts the producers based on this priority value.
- If no control strategies are assigned to the producers, it just uses the already existing marginal cost sorting system.
- If the priority control strategy is in place, and one or more of the producers do not have a priority value assigned, they received the lowest possible priority value. In this case, a warning message is returned to the user.
- The controller sorts the producers with unique priority values. If more than one producer has the same priority value assigned, the controller will sort them sequentially, so they all end up having a unique priority value. This way, least one of the producers is re-assigned a priority value lower than the one it had originally assigned, in order to break the "priority tie".

### CODE REVIEW

File: src/omotes_simulator_core/solver/network/network.py [hunk 1]
```diff
@@ -20,10 +20,12 @@
 from omotes_simulator_core.solver.network.assets.base_asset import BaseAsset
 from omotes_simulator_core.solver.network.assets.boundary import BaseBoundary
 from omotes_simulator_core.solver.network.assets.fall_type import FallType
+from omotes_simulator_core.solver.network.assets.heat_transfer_asset import (
+    HeatTransferAsset,
+)
 from omotes_simulator_core.solver.network.assets.node import Node
```
Comment (MichielTukker) [inline src/omotes_simulator_core/solver/network/network.py:26]: is this relevant for this PR?

File: src/omotes_simulator_core/adapter/transforms/controller_mappers/controller_producer_mapper.py [hunk 2]
```diff
@@ -42,12 +42,14 @@ def to_entity(self, esdl_asset: EsdlAssetObject) -> ControllerProducer:
...
         temperature_out = esdl_asset.get_temperature("Out", "Supply")
+        strategy_priority = esdl_asset.get_strategy_priority()
         contr_producer = ControllerProducer(
             name=esdl_asset.esdl_asset.name,
             identifier=esdl_asset.esdl_asset.id,
             temperature_in=temperature_in,
             temperature_out=temperature_out,
             power=power,
             marginal_costs=marginal_costs,
+            priority=strategy_priority,
```
Comment (MichielTukker) [inline src/omotes_simulator_core/adapter/transforms/controller_mappers/controller_producer_mapper.py:53]: `strategy_priority` may be `None`,  while ControllerProducer expects an int.   We should (at least) also modify the ControllerProducer  type hints

**Review by MichielTukker — CHANGES_REQUESTED** [review]
See comments

Please also add a unit test to ensure that the `priority = None`  doesn't cause issues for ControllerProducer

File: src/omotes_simulator_core/entities/assets/esdl_asset_object.py [hunk 3]
```diff
@@ -55,6 +55,13 @@ def get_id(self) -> str:
         """Get the id of the asset."""
         return str(self.esdl_asset.id)
 
+    def get_strategy_priority(self) -> Any:
```
[the remark this answers is not in the document]
Comment (Jaimepatt) [reply src/omotes_simulator_core/entities/assets/esdl_asset_object.py:58]: I changed it, but now the type checker seems to be broken. It gives and error for int | Any even for files I didn't touch in this PR.

File: src/omotes_simulator_core/solver/network/network.py (hunk 1 above)
Comment (Jaimepatt) [reply src/omotes_simulator_core/solver/network/network.py:26]: This is indeed not part of my PR, but it gets flagged by the linter. It won't pass the PR test unless I add this change.

File: src/omotes_simulator_core/adapter/transforms/controller_mappers/controller_producer_mapper.py (hunk 2 above)
Comment (Jaimepatt) [reply src/omotes_simulator_core/adapter/transforms/controller_mappers/controller_producer_mapper.py:53]: Fixed by changing the hints.

**Comment by Jaimepatt** [comment]
@MichielTukker as per your comment, I added an extra test to test_controller_producer.py.

**Review by MichielTukker — CHANGES_REQUESTED** [review]
can we rebase before merge?

File: src/omotes_simulator_core/solver/network/network.py (hunk 1 above)
Comment (Jaimepatt) [reply src/omotes_simulator_core/solver/network/network.py:26]: Marged main into this branch.

File: src/omotes_simulator_core/entities/network_controller.py [hunk 4]
```diff
@@ -60,7 +64,20 @@ def __init__(
...
+            self._set_priority_from_marginal_costs()
+
+    def _check_strategy_priority(self) -> bool:
+        """Check if at least one asset has a control strategy priority assigned."""
+        strategy_flag = False
+        for producer in self.producers:
+            if producer.priority is not None:
+                strategy_flag = True
+                break
+        return strategy_flag
```
[the remark this answers is not in the document]
Comment (Jaimepatt) [reply src/omotes_simulator_core/entities/network_controller.py]: I merged the latest version of main into this branch. Also converted the mentioned function into a single line.

**Review by MichielTukker — APPROVED** [review]
